### ISSUE TITLE
Type resolution: expect and actual from different KMP source sets are analysed as one fragment

### ISSUE THREAD

**Issue by dc-abuettgenbach** [detekt/detekt#9602]
## Expected Behavior

A Kotlin Multiplatform module whose `expect` declaration lives in `commonMain` and whose `actual`
lives in a platform source set should analyse without compiler errors, the same way it compiles.

## Observed Behavior

Every type-resolution task that spans two source sets reports three errors per `expect`/`actual`
pair. Minimal project, one `expect` and one `actual`:

```
src/commonMain/kotlin/Sample.kt:1:1: error: defaultLanguage: expect and corresponding actual are declared in the same module.
src/jvmMain/kotlin/Sample.jvm.kt:1:1: error: 'internal final actual fun defaultLanguage(): String' has no corresponding expected declaration
src/jvmMain/kotlin/Sample.jvm.kt:1:1: error: defaultLanguage: expect and corresponding actual are declared in the same module.
There were 3 compiler errors found during analysis. This affects accuracy of reporting.
```

`./gradlew compileKotlinJvm` on the very same sources is `BUILD SUCCESSFUL`, so the code is valid —
only the analysis rejects it.

It scales linearly with the number of `expect` declarations. In our build, seven modules produce
exactly `3 × (number of top-level expect declarations)`: 3 errors where there is one `expect`, 18
where there are six, 24 where there are eight. That arithmetic accounts for every one of them with
no residue, which is how we are confident this is a single cause rather than several.

Note the task still reports `BUILD SUCCESSFUL` — the analysis degrades silently. Type-aware rules
then misfire in both directions. We have confirmed three false `UnusedPrivateProperty` findings in
these modules, each disproved by finding the call sites by hand, and they were nearly acted on as
dead code.

## Steps to Reproduce

JVM-only, no Android SDK needed. Four files:

`settings.gradle.kts`
```kotlin
pluginManagement { repositories { gradlePluginPortal(); mavenCentral() } }
dependencyResolutionManagement { repositories { mavenCentral() } }
rootProject.name = "detekt-kmp-expect-actual"
```

`build.gradle.kts`
```kotlin
plugins {
    kotlin("multiplatform") version "2.4.10"
    id("dev.detekt") version "2.0.0-alpha.6"
}
repositories { mavenCentral() }
kotlin { jvm() }
detekt { debug = true } // needed to see the messages; `debug` is @Console, so it is not a task input
```

`src/commonMain/kotlin/Sample.kt`
```kotlin
internal expect fun defaultLanguage(): String
```

`src/jvmMain/kotlin/Sample.jvm.kt`
```kotlin
internal actual fun defaultLanguage(): String = "en"
```

```
./gradlew compileKotlinJvm          # BUILD SUCCESSFUL
./gradlew detektMainJvm --rerun     # the three errors above
```

`--rerun` matters: `debug` is annotated `@get:Console`, so toggling it does not invalidate the task
and a cached run prints nothing.

The same thing happens on `detektMainAndroid` with `com.android.library` and the `actual` in
`androidMain`.

## Context

[#7533](https://github.com/detekt/detekt/issues/7533) was closed as fixed by passing
`-Xmulti-platform`, and the Gradle plugin does pass it (`MultiPlatformEnabledArgument` in
`Detekt.kt`). This looks like the next problem behind that flag rather than a regression of it: the
flag enables multiplatform *language* support, but the analysis still receives every source set of
the compilation as one flat `--input` list, so `commonMain` and `jvmMain` land in a single
fragment — which is the arrangement `EXPECT_AND_ACTUAL_IN_THE_SAME_MODULE` exists to reject.

The Kotlin Gradle plugin avoids this not by compiling the source sets separately but by declaring
them: KGP 2.4 passes `-Xfragments`, `-Xfragment-sources` and `-Xfragment-refines` to the same
frontend, which is what puts the `expect` and the `actual` in different fragments of one module.
detekt's task builds `--input` from the compilation's source sets with no fragment structure
attached.

Filing this because [#9293](https://github.com/detekt/detekt/issues/9293) records that the
"compiler errors found during analysis" warning is kept deliberately, since it marks a real defect
worth having reproducers for. This is one such defect.

Deliberately not filed as a duplicate of
[#7531](https://github.com/detekt/detekt/issues/7531): that one is about compiler plugins
(kotlinx.serialization) and its author has kept it scoped there. Our build hits both, separately.

## Your Environment

* Version of detekt used: `2.0.0-alpha.6`
* Version of Gradle used (if applicable): `9.7.0`
* Gradle scan link: not available (offline build host)
* Operating System and version: reproduced on macOS 26.5.2 (arm64, JDK 21.0.11); the same three
  errors appear on Linux
* Kotlin: `2.4.10`
* Link to your project: private

**Comment by Munzey** [comment]
We’re hitting this too on detekt `2.0.0-alpha.6` with Kotlin `2.4.10`. A KMP module with an `expect` in `commonMain` and its `actual` in `jvmMain` produces 
>There were 3 compiler errors found during analysis

 on the JVM type-resolution task (`detektMainJvm`).
